openmls/src/treesync/mod.rs

```diff
@@ -482,6 +488,31 @@ impl TreeSync {
                     TreeSyncFromNodesError::from(PublicTreeError::InvalidParentHash)
                 }
             })?;
+
+        // Check that no two nodes share an encryption key.
+        // This is a bit stronger than what the spec requires: It requires that the encryption keys
+        // in parent nodes and unmerged leaves must be unique. Here, we check that all encryption
+        // keys (all leaf nodes, incl. unmerged and all parent nodes) are unique.
+        //
+        // https://validation.openmls.tech/#valn1410
+        tree_sync
+            .tree
+            .leaves()
+            .flat_map(|(_, node)| node.node().as_ref().map(LeafNode::encryption_key))
+            .chain(
+                tree_sync
+                    .tree
+                    .parents()
+                    .flat_map(|(_, node)| node.node().as_ref().map(ParentNode::encryption_key)),
+            )
+            .try_fold(HashSet::new(), |mut leaves, key| {
+                if !leaves.insert(key) {
+                    return Err(RatchetTreeError::DuplicateEncryptionKey);
+                }
+
+                Ok(leaves)
+            })?;
```

Comment:
This probably isn't that much of a performance concern (and we can do it in a separate PR), but between the different validation steps we iterate over the whole tree multiple times. I wonder if we can't streamline that somehow.

Reply:
I pasted the body of the loop that iterated over both leaves and parents into each of the two loops in that I do in the other file. There might be more to gain somewhere else, but there was nothing that I saw that would be obvious and wouldn't clutter the code.

Reply:
Thanks!